src/Raven.Server/Documents/PeriodicBackup/DirectUpload/DirectFileUploader.cs

```diff
@@ -0,0 +1,199 @@
+        {
+            ClientFactory = clientFactory,
+            Key = CombinePathAndKey(remoteFolderName, folderName, fileName),
+            Metadata = new Dictionary<string, string>
+            {
+                { "Description", GetBackupDescription() }
+            },
+            IsFullBackup = _isFullBackup,
+            RetentionPolicyParameters = _retentionPolicyParameters,
+            CloudUploadStatus = _backupResult.S3Backup,
+            OnBackupException = _settings.OnBackupException,
+            OnProgress = AddInfo
+        };
+    }
+
+    public override string GetBackupDescription()
+    {
+        return $"{nameof(DirectFileUploader)}";
+    }
+
+    public void AddDelete(string folderName, string fileName)
+    {
+        switch (_destination)
+        {
+            case BackupConfiguration.BackupDestination.AmazonS3:
+                CreateLongRunningDeleteThread(_settings.S3Settings, DeleteFromS3, S3Name, folderName, fileName, 
+                    ThreadNames.ForDeleteRetiredAttachment(ThreadName(S3Name), _settings.DatabaseName, S3Name, _settings.TaskName));
+                break;
+
+            case BackupConfiguration.BackupDestination.Azure:
+                var threadInfo = ThreadNames.ForDeleteRetiredAttachment(ThreadName(AzureName), _settings.DatabaseName, AzureName, _settings.TaskName);
+                CreateLongRunningDeleteThread(_settings.AzureSettings, DeleteFromAzure, AzureName, folderName, fileName, threadInfo);
+                break;
+
+            default:
+                throw new ArgumentOutOfRangeException($"Missing implementation for direct upload destination '{_destination}'");
+        }
+    }
+
+    public IDictionary<string, string> GetObjectMetadata(string folderName, string objKeyName)
+    {
+        switch (_destination)
+        {
+            case BackupConfiguration.BackupDestination.AmazonS3:
+                return GetObjectMetadataFromS3(_settings.S3Settings, folderName, objKeyName);
+
+            case BackupConfiguration.BackupDestination.Azure:
+                return GetObjectMetadataFromAzure(_settings.AzureSettings, folderName, objKeyName);
+
+            default:
+                throw new ArgumentOutOfRangeException($"Missing implementation for direct upload destination '{_destination}'");
+        }
+    }
+
+    private string ThreadName(string name)
+    {
+        return $"Delete retired attachment of database '{_settings.DatabaseName}' from {name} (task: '{_settings.TaskName}')";
+    }
+
+    public void CreateUploadTask(DocumentDatabase database, Stream attachmentStream, string objKeyName, CancellationToken token)
+    {
+        var targetName = _destination == BackupConfiguration.BackupDestination.AmazonS3 ? S3Name : AzureName;
+        var threadName = $"Upload retired attachment '{objKeyName}' of database '{_settings.DatabaseName}' from {targetName} (task: '{_settings.TaskName}')";
+        var thread = PoolOfThreads.GlobalRavenThreadPool.LongRunning(_ =>
```

Comment:
Why is that a _long running_ thread? What isn't that a task?

Reply:
I used the same logic as in `BackupUploader.cs` since we also might have a huge attachment: 
https://github.com/garayx/ravendb/blob/d1bb408c06a442fa9456aefbcb97c7f870cd1e4e/src/Raven.Server/Documents/PeriodicBackup/BackupUploader.cs#L178

Reply:
That is a different situation = for backups - we have a single large task, that is very rarely executed.

Here we can assume that we have a _lot_ of them, and that they are likely small - tasks are better here.
Not to mention the complexity of managing the concurrency here with the limit of concurrent uploads